waiter/src/waiter/scheduler/composite.clj

```diff
@@ -0,0 +1,165 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns waiter.scheduler.composite
+  (:require [clojure.tools.logging :as log]
+            [plumbing.core :as pc]
+            [schema.core :as s]
+            [waiter.scheduler :as scheduler]
+            [waiter.util.utils :as utils]))
+
+(defn process-invalid-services
+  "Deletes the provided services on the specified scheduler."
+  [scheduler service-ids]
+  (log/info "found" (count service-ids) "misplaced services in" scheduler)
+  (doseq [service-id service-ids]
+    (log/info "deleting misplaced service" service-id "in" scheduler)
+    (scheduler/delete-service scheduler service-id))
+  (log/info "deleted" (count service-ids) "misplaced services in" scheduler))
+
+(defn retrieve-service->instances
+  "Retrieves the service->instances for services that are configured to be running on the specified scheduler.
+   If any service is found (e.g. due to default scheduler being changed) which does not belong in the
+   specified scheduler, it is promptly deleted."
+  [scheduler service-id->scheduler]
+  (let [service->instances (scheduler/get-service->instances scheduler)
+        valid-service? #(-> % key :id service-id->scheduler (= scheduler))
+        {valid-service->instances true invalid-service->instances false} (group-by valid-service? service->instances)]
+    (when (seq invalid-service->instances)
+      (->> invalid-service->instances
+           keys
+           (map :id)
+           (process-invalid-services scheduler)))
+    (into {} valid-service->instances)))
+
+(defn retrieve-services
+  "Retrieves the services for services that are configured to be running on the specified scheduler.
+   If any service is found (e.g. due to default scheduler being changed) which does not belong in the
+   specified scheduler, it is promptly deleted."
+  [scheduler service-id->scheduler]
+  (let [services (scheduler/get-services scheduler)
+        valid-service? #(-> % :id service-id->scheduler (= scheduler))
+        {valid-services true invalid-services false} (group-by valid-service? services)]
+    (when (seq invalid-services)
+      (->> invalid-services
+           (map :id)
+           (process-invalid-services scheduler)))
+    (into [] valid-services)))
+
+(defrecord CompositeScheduler [service-id->scheduler scheduler-id->scheduler]
```

Comment:
We need to handle unavailability of component schedulers.

Reply:
`service-id->scheduler` throws an exception if it can't find a scheduler for a `service-id`.

Reply:
What I mean is, what if one of the schedulers is down while the other is still available?